File: programmers/소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> check, visit;
string num, temp;
int answer = 0, N;

void solve() {
    for (int index = 0; index < N; index++) {
        if (visit[index] == 0) {
            visit[index] = 1;
            temp.push_back(num[index]);
            if (check[stoi(temp)] == 0 && stoi(temp) >= 2) {
                check[stoi(temp)] = 1;
                answer++;
            }
            solve();
            visit[index] = 0;
            temp.pop_back();
        }
    }
}

int solution(string numbers) {
    vector<int> number;
    N = numbers.size();
    visit.assign(N, 0);
    num = numbers;
    int target = 0, times = 1;
    for (int index = 0; index < N; index++) {
        number.push_back(numbers[index] - '0');
    }
    sort(number.begin(), number.end());
    for (int index = 0; index < N; index++) {
        target += number[index] * times;
        times *= 10;
    }
    check.assign(target + 1, 0);
    for (int index = 2; index <= target; index++) {
        for (int pos = 2; index * pos <= target; pos++) {
            if (check[index * pos] == 1) continue;
            check[index * pos] = 1;
        }
    }
    solve();
    return answer;
}

File: programmers/소수 찾기/소수 찾기.kt
import java.util.*
class Solution {
    fun solution(numbers: String): Int {
        val size = numbers.length
        val check = BooleanArray(size)
        val tree = TreeSet<Int>()
        
        fun search(cnt: Int, number: String) {
            if ((cnt <= size) and (cnt > 0)) {
                tree.add(number.toInt())
            }
            if (cnt == size) return
            
            repeat(size) loop@ {
                if (check[it]) return@loop
                check[it] = true
                search(cnt + 1, number + numbers[it])
                check[it] = false
            }
        }
        search(0, "")
        
        fun isSosu(target: Int): Boolean {
            if (target < 2) return false
            var num = 2
            while ((num * num) <= target) {
                if (target % num == 0) return false
                num++
            }
            return true
        }
        
        var answer = 0
        for (num in tree) {
            if (isSosu(num)) answer++
        }
        
        return answer
    }
}